pkcs10/src/attribute.rs

```diff
@@ -0,0 +1,40 @@
+use der::asn1::{Any, ObjectIdentifier, SetOfVec};
+use der::{Decodable, OrdIsValueOrd, Sequence};
+
+/// PKCS#10 `Attribute` as defined in [RFC 2986 Section 4].
+///
+/// ```text
+/// Attribute { ATTRIBUTE:IOSet } ::= SEQUENCE {
+///     type   ATTRIBUTE.&id({IOSet}),
+///     values SET SIZE(1..MAX) OF ATTRIBUTE.&Type({IOSet}{@type})
+/// }
+/// ```
+///
+/// [RFC 2986 Section 4]: https://datatracker.ietf.org/doc/html/rfc2986#section-4
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Sequence)]
+pub struct Attribute<'a> {
```

Comment:
As a bit of a sidebar, it'd be nice if we could find somewhere more central to define `Attribute`, possibly in the `spki` crate.

Reply:
@tarcieri I agree. I was hoping to get this merged and then have a conversation about how to structure the type definitions.

Basically, we have the attribute types, including `Attribute`, `Attributes`, `AttributeType`, `AttributeValue`, and `AttributeTypeAndValue` (the last one is currently in the x509 crate). We also have the X.501 name types along with common name definitions (currently in the `x509` crate). We also have the `Extension` type along with common extensions (in the `x509` crate). We also have the `spki` crate with its types.

All of these types are shared across different document types and could live in a shared crate (`x509-core`?).

Reply:
They are not all the same so will need to be careful. I agree we could collapse where possible (but might want to leave a pointer in module that is losing a definition).

Reply:
@carl-wallace They are actually the same. The history of these types is that early definitions used advanced features of ASN.1 to constrain the types. But these features turned out to be too difficult to implement in the actual ASN.1 parsers. So over time they relaxed the constraints. We actually implement the later, unconstrained, version as defined by RFC 5280:

```
Attribute               ::= SEQUENCE {
      type             AttributeType,
      values    SET OF AttributeValue }
            -- at least one value is required

AttributeType           ::= OBJECT IDENTIFIER

AttributeValue          ::= ANY -- DEFINED BY AttributeType

AttributeTypeAndValue   ::= SEQUENCE {
        type    AttributeType,
        value   AttributeValue }
```

This is the version that should be implemented everywhere.

Reply:
You have the history backwards, they were originally defined without the constraints (in 1988 ASN.1 syntax) then later updated (to the 2002 syntax mostly), however, the bits on the wire did not change in that exchange. 

My point is different but your snip illustrates it. The Attribute definition in PCKS10 features a values field with a SET OF type. The AttributeTypeAndValue definition features a value field that is not a SET OF (or at least is not necessarily a SET OF). These are different types. Sometimes these definitions are repeated in modules vs. importing, leaving a mess.

If your point is that we could likely collapse all references down to the definitions above, I probably agree with you but would want to check source modules before each referencing. I mostly favor organizing by ASN.1 module so that the imports flow in the same way as the specs (to reduce confusion). It's probably too late in the game for that though. It's not a big deal either way, we just need to make sure the types are as they need to be.

Reply:
@carl-wallace I don't think anyone here is confusing the `Attribute` type with the `AttributeTypeAndValue` type.